zerver/views/auth.py

```diff
@@ -437,23 +470,28 @@ def authenticate_remote_user(realm: Realm,
     return user_profile
 
 _subdomain_token_salt = 'zerver.views.auth.log_into_subdomain'
+LOGIN_KEY_PREFIX = "login_key_"
+LOGIN_KEY_FORMAT = LOGIN_KEY_PREFIX + "{token}"
+LOGIN_KEY_EXPIRATION_SECONDS = 15
+LOGIN_TOKEN_LENGTH = UserProfile.API_KEY_LENGTH
 
 @log_view_func
 def log_into_subdomain(request: HttpRequest, token: str) -> HttpResponse:
-    """Given a valid signed authentication token (generated by
+    """Given a valid authentication token (generated by
     redirect_and_log_into_subdomain called on auth.zulip.example.com),
     call login_or_register_remote_user, passing all the authentication
-    result data that had been encoded in the signed token.
+    result data that has been stored in redis, associated with this token.
+    Obligatory fields for the data are 'subdomain' and 'email', because this endpoint
+    needs to know which user and realm to log into. Others are optional and only used
+    if the user account still needs to be made and they're passed as argument to the
+    register_remote_user function.
     """
-
-    try:
-        data = signing.loads(token, salt=_subdomain_token_salt, max_age=15)
-    except signing.SignatureExpired as e:
-        logging.warning('Subdomain cookie: {}'.format(e))
-        return HttpResponse(status=400)
-    except signing.BadSignature:
-        logging.warning('Subdomain cookie: Bad signature.')
-        return HttpResponse(status=400)
+    if has_api_key_format(token):  # The tokens are intended to have the same format as API keys.
+        data = get_login_data(token)
+        if data is None:
+            raise JsonableError(_("Invalid one-time authentication key"))
+    else:
+        raise JsonableError(_("Malformed one-time authentication key"))
```

Comment:
Does it make sense to convert these into JsonableErrors?  This is a main browser window route, so I don't think it really makes sense (that's why the previous code just returned a generic Django 400 page, rather than a JSON-format error).  

I also don't think we need to provide two different translated error messages here (or even 1) -- this is a corner case that won't come up unless someone's trying to hack Zulip.

Finally, we should do early-return with `if not has_api_key_format...` for more readable code.

Reply:
Makes sense, changed the block to:
```
    if not has_api_key_format(token):  # The tokens are intended to have the same format as API keys.
        logging.warning("log_into_subdomain: Malformed token given: %s" % (token,))
        return HttpResponse(status=400)

    data = get_login_data(token)
    if data is None:
        logging.warning("log_into_subdomain: Invalid token given: %s" % (token,))
        return HttpResponse(status=400)
```
(and adjusted tests)